File: PrivateVault.ai-main/sdk/node/src/approvals.js
import crypto from 'crypto';

export function createApproval({
  approverId,
  role,
  region,
  intentHash,
  keyId,
  secret,
  issuedAt,
  expiresAt,
  approvalId,
}) {
  const issued = issuedAt || Math.floor(Date.now() / 1000);
  const approval = {
    approval_id: approvalId || null,
    approver_id: approverId,
    role,
    region,
    intent_hash: intentHash,
    issued_at: issued,
    expires_at: expiresAt || issued + 3600,
    key_id: keyId,
  };
  const signature = crypto.createHmac('sha256', secret).update(intentHash).digest('hex');
  approval.signature = signature;
  return approval;
}

File: PrivateVault.ai-main/sdk/python/privatevault_sdk/approvals.py
import hashlib
import hmac
import time
from typing import Dict


def create_approval(
    *,
    approver_id: str,
    role: str,
    region: str,
    intent_hash: str,
    key_id: str,
    secret: str,
    issued_at: int | None = None,
    expires_at: int | None = None,
    approval_id: str | None = None,
) -> Dict:
    issued = issued_at or int(time.time())
    approval = {
        "approval_id": approval_id,
        "approver_id": approver_id,
        "role": role,
        "region": region,
        "intent_hash": intent_hash,
        "issued_at": issued,
        "expires_at": expires_at or (issued + 3600),
        "key_id": key_id,
    }
    signature = hmac.new(secret.encode("utf-8"), intent_hash.encode("utf-8"), hashlib.sha256).hexdigest()
    approval["signature"] = signature
    return approval
